From 456f263be4ceb630b45ef1c52e72a06c86d55c2d Mon Sep 17 00:00:00 2001
From: Brett Mastbergen <bmastbergen@ciq.com>
Date: Wed, 9 Apr 2025 09:15:21 -0400
Subject: [PATCH] net/sched: cls_route: No longer copy tcf_result on update to
 avoid use-after-free

jira VULN-6649
cve CVE-2023-4206
commit-author valis <sec@valis.email>
commit b80b829e9e2c1b3f7aae34855e04d8f6ecaf13c8
rebase-comment: CVE-2023-4218 was droped as it was a typo for an eclipse
	CVE, and the intended one CVE-2023-4128 was rejected in favor of
	CVE-2023-4206, CVE-2023-4207, CVE-2023-4208.

When route4_change() is called on an existing filter, the whole
tcf_result struct is always copied into the new instance of the filter.

This causes a problem when updating a filter bound to a class,
as tcf_unbind_filter() is always called on the old instance in the
success path, decreasing filter_cnt of the still referenced class
and allowing it to be deleted, leading to a use-after-free.

Fix this by no longer copying the tcf_result struct from the old filter.

Fixes: 1109c00547fc ("net: sched: RCU cls_route")
	Reported-by: valis <sec@valis.email>
	Reported-by: Bing-Jhong Billy Jheng <billy@starlabs.sg>
	Signed-off-by: valis <sec@valis.email>
	Signed-off-by: Jamal Hadi Salim <jhs@mojatatu.com>
	Reviewed-by: Victor Nogueira <victor@mojatatu.com>
	Reviewed-by: Pedro Tammela <pctammela@mojatatu.com>
	Reviewed-by: M A Ramdhan <ramdhan@starlabs.sg>
Link: https://lore.kernel.org/r/20230729123202.72406-4-jhs@mojatatu.com
	Signed-off-by: Jakub Kicinski <kuba@kernel.org>
(cherry picked from commit b80b829e9e2c1b3f7aae34855e04d8f6ecaf13c8)
	Signed-off-by: Brett Mastbergen <bmastbergen@ciq.com>
---
 net/sched/cls_route.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/net/sched/cls_route.c b/net/sched/cls_route.c
index 48712bc51bda..194468d0355a 100644
--- a/net/sched/cls_route.c
+++ b/net/sched/cls_route.c
@@ -511,7 +511,6 @@ static int route4_change(struct net *net, struct sk_buff *in_skb,
 	if (fold) {
 		f->id = fold->id;
 		f->iif = fold->iif;
-		f->res = fold->res;
 		f->handle = fold->handle;
 
 		f->tp = fold->tp;
-- 
2.47.3

